src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -1057,13 +1117,12 @@ private MethodSymbol PerformForEachPatternOverloadResolution(TypeSymbol patternT
             {
                 result = overloadResolutionResult.ValidResult.Member;
 
-                if (result.IsStatic || result.DeclaredAccessibility != Accessibility.Public)
+                if (result.IsStatic && !viaExtensionMethod || result.DeclaredAccessibility != Accessibility.Public)
                 {
-                    if (warningsOnly)
-                    {
-                        MessageID patternName = isAsync ? MessageID.IDS_FeatureAsyncStreams : MessageID.IDS_Collection;
-                        diagnostics.Add(ErrorCode.WRN_PatternStaticOrInaccessible, _syntax.Expression.Location, patternType, patternName.Localize(), result);
-                    }
+                    var patternName = isAsync ? MessageID.IDS_FeatureAsyncStreams : MessageID.IDS_Collection;
+                    var errorCode = warningsOnly ? ErrorCode.WRN_PatternInaccessibleOrNotInstance : ErrorCode.ERR_PatternInaccessibleOrNotInstance;
```

Comment:
As another example of the complicated logic here: why aren't we doing the same thing for both extension and regular methods? I would expect us to warn for this, and then report `ERR_ForEachMissingMember`, just like regular `foreach` methods, or for there to be just one shared error between the two.

Reply:
That's what I did originally, but I thought you asked me to change it:

https://github.com/dotnet/roslyn/pull/43050#discussion_r405021098

Reply:
Hmm. I think you were right about the noise. Sorry to ask, but can you revert it?

---
In reply to: [416819889](https://github.com/dotnet/roslyn/pull/43050#discussion_r416819889) [](ancestors = 416819889)

Reply:
Yeah, this does look much better, even if we have multiple warnings. Thanks.

---
In reply to: [416822495](https://github.com/dotnet/roslyn/pull/43050#discussion_r416822495) [](ancestors = 416822495,416819889)